### DISCUSSION TITLE
Critical CSS

### DISCUSSION THREAD

**Discussion by childrentime** [vercel/next.js#59402] [Help]
### Summary

 I'm using CSS modules with Next.js to style my application. However, I haven't found a way to handle critical CSS. There's no hook similar to useStyles(styles) to collect styles during the initial server-side rendering. I don't want to bundle all the CSS files as external stylesheets; instead, I want to convert the CSS imported through useStyles during the rendering process into inline style tags and insert them into the HTML.

**Comment by matthewraaff** [comment]
Hi, there are some libraries such as [gulp-inline-css](https://www.npmjs.com/package/gulp-inline-css) that allow you to do this.

The package is about 38 LOC, feel free to have a read through and reimplement it if you have any specific requirements.

**Reply by childrentime** [reply]
I need a way to collect css in server side at initial render, it is not static.

**Comment by Milkywayrules** [comment]
is Nextjs not automatically handle that critical CSS for us? by splitting the stylesheet into CSS Module, I believe they'll lazyload the style (same behvior with the jsx file itself), not loading all at once (except we write it into global.css)

CMIIW

**Comment by childrentime** [comment]
@Milkywayrules Next.js processes CSS modules into external CSS files and includes them in the project. If your project has many conditions for using certain modules, such as {ABTEST ? <A/>: <B/>}, the CSS from both A and B will be packaged into the CSS files. The size of the CSS file increases with the number of modules, as it does not only include the essential CSS.

**Comment by erdalugur** [comment]
same !!

**Comment by jahilldev** [comment]
Next supports an experimental `optimizeCss` flag that you can try out in your config, e.g:
```js
// next.config.js

experimental: {
  optimizeCss: true, // also accepts a config object, see below
}
```

Configuration options:
https://github.com/GoogleChromeLabs/critters/blob/a590c05f9197b656d2aeaae9369df2483c26b072/packages/critters/src/index.d.ts#L41-L60

I haven't found anywhere that officially documents this, aside from a tweet and some Github comments:
https://twitter.com/hdjirdeh/status/1369709676271726599
https://github.com/vercel/vercel/discussions/5973#discussioncomment-472618

In my initial tests this works well (production build locally, haven't tried deploying yet) with Tailwind. Not sure how it'll play with CSS Modules or other solutions like CSS in JS, but worth a try.

The performance benefits on inline critical CSS are substantial 👍

**Comment by childrentime** [comment]
well. this doesn't work with css module. It did'nt reduce unused css

**Comment by childrentime** [comment]
https://github.com/childrentime/my-app

**Reply by Mifaresss** [reply]
Critters don't compatible with app directory (it isn't working)
